### PR TITLE
Add write method to BoutOptionsFile and update showdata

### PR DESCRIPTION
**Pull request by johnomotani**
I've pulled these changes out of #1106 because they are simpler. Adding write() to BoutOptionsFile is backward compatible, and the t_array change to showdata may change some animation titles but should not break anything.

### CODE REVIEW

File: tools/pylib/boutdata/data.py [hunk 1]
```diff
@@ -247,6 +248,44 @@ def __init__(self, filename, name="root"):
...
+            (defaults to path of the file that was read in)
+        force_overwrite : bool
+            If False then prompt the user whether to overwrite an existing
+            file, if there is one. Otherwise, just overwrite without asking.
+            (default False)
+        """
+        if filename is None:
+            filename = self.filename
+
+        if not force_overwrite and os.path.exists(filename):
```
Comment (bendudson) [inline tools/pylib/boutdata/data.py]: `force_overwrite` should be `overwrite` to be consistent with arguments.

File: tools/pylib/boutdata/data.py [hunk 2]
```diff
@@ -247,6 +248,44 @@ def __init__(self, filename, name="root"):
...
+        user for confirmation.
+        Note, no comments from the original input file are transferred to the
+        new one.
+
+        Parameters
+        ----------
+        filename : str
+            Path of the file to write
+            (defaults to path of the file that was read in)
+        force_overwrite : bool
```
Comment (bendudson) [inline tools/pylib/boutdata/data.py]: Comments should be updated: user isn't prompted anymore but an exception is raised

Comment (johnomotani) [reply tools/pylib/boutdata/data.py]: This and also the previous comment `force_overwrite`->`overwrite` are fixed now.

**Comment by ZedThree** [comment]
This does change the previous behaviour for `showdata`, though it doesn't look like it breaks anything. Does anyone have strong opinions one or the other? I do see the appeal of not calling `collect` in a plotting routine.

File: tools/pylib/boutdata/data.py [hunk 3]
```diff
@@ -247,6 +248,44 @@ def __init__(self, filename, name="root"):
...
+            (default False)
+        """
+        if filename is None:
+            filename = self.filename
+
+        if not overwrite and os.path.exists(filename):
+            raise ValueError("Not overwriting existing file, cannot write output to "+filename)
+
+        def write_section(basename, opts, f):
+            if basename is not "":
```
Comment (ZedThree) [inline tools/pylib/boutdata/data.py]: Could just be `if basename:`, unless I'm missing something? Similarly for the same conditional below.

File: tools/pylib/boutdata/data.py [hunk 4]
```diff
@@ -247,6 +248,44 @@ def __init__(self, filename, name="root"):
...
+                f.write("["+basename+"]\n")
+            for key, value in opts._keys.items():
+                f.write(key+" = "+str(value)+"\n")
+            for section in opts.sections():
+                if basename is not "":
+                    write_section(basename+":"+section, opts[section], f)
+                else:
+                    write_section(section, opts[section], f)
```
Comment (ZedThree) [inline tools/pylib/boutdata/data.py]: I would be tempted to write this to remove the semi-duplicated branch:

```python
section_name = basename+":"+section if basename else section
write_section(section_name, opts[section, f)
```
although I appreciate that the ternary can be a bit confusing

**Pull request merged**